tests/unit/test_lidar_parser.py

```diff
@@ -0,0 +1,207 @@
+    """
+    Fixture to initialize a LidarParser instance for testing.
+    """
+    parser = lidar_parser.LidarParser()
+    yield parser
+
+
+@pytest.fixture()
+def lidar_detection_up() -> lidar_detection.LidarDetection:  # type: ignore
+    """
+    Fixture to create an upward LidarDetection object.
+    """
+    result, detection = lidar_detection.LidarDetection.create(5.0, ANGLE_UP)
+    assert result
+    assert detection is not None
+    yield detection
+
+
+@pytest.fixture()
+def higher_angle_detection_up(lidar_detection_up: lidar_detection.LidarDetection) -> lidar_detection.LidarDetection:  # type: ignore
+    """
+    Fixture to create a LidarDetection object with a higher upward angle than the previous detection.
+    """
+    result, higher_detection = lidar_detection.LidarDetection.create(
+        5.0, lidar_detection_up.angle + 5.0
+    )
+    assert result
+    assert higher_detection is not None
+    yield higher_detection
+
+
+@pytest.fixture()
+def lidar_detection_down() -> lidar_detection.LidarDetection:  # type: ignore
+    """
+    Fixture to create a downward LidarDetection object.
+    """
+    result, detection = lidar_detection.LidarDetection.create(5.0, ANGLE_DOWN)
+    assert result
+    assert detection is not None
+    yield detection
+
+
+@pytest.fixture()
+def lower_angle_detection_down(lidar_detection_up: lidar_detection.LidarDetection) -> lidar_detection.LidarDetection:  # type: ignore
+    """
+    Fixture to create a LidarDetection object with a lower downward angle than the previous detection.
+    """
+    result, lower_detection = lidar_detection.LidarDetection.create(
+        5.0, lidar_detection_up.angle - 5.0
+    )
+    assert result
+    assert lower_detection is not None
+    yield lower_detection
+
+
+class TestLidarParser:
+    """
+    Tests for the LidarParser run() method.
+    """
+
+    def test_initial_run_no_oscillation(
+        self,
+        lidar_parser_instance: lidar_parser.LidarParser,
+        lidar_detection_up: lidar_detection.LidarDetection,
+    ) -> None:
+        """
+        Test that a single upward detection does not complete an oscillation.
+        """
+        result, oscillation = lidar_parser_instance.run(lidar_detection_up)
+        assert not result
+        assert oscillation is None
+
+    def test_oscillation_detected_up_to_down(
+        self,
+        lidar_parser_instance: lidar_parser.LidarParser,
+        lidar_detection_up: lidar_detection.LidarDetection,
+        higher_angle_detection_up: lidar_detection.LidarDetection,
+        lidar_detection_down: lidar_detection.LidarDetection,
+    ) -> None:
+        """
+        Test detection of an oscillation with a change in direction from up to down after multiple detections.
+        """
+        lidar_parser_instance.run(lidar_detection_up)
+        lidar_parser_instance.run(higher_angle_detection_up)
+        result, oscillation = lidar_parser_instance.run(lidar_detection_down)
+        assert result
```

Comment:
check list length (i.e. len == 2) and check whether the oscillation contains only the readings going up i.e. it sees 15 and 20 degrees. (lidar_oscillation.readings[0].angle == 15)

Reply:
1. expected = lidar_detection.create(5, 10)
2. oscillation = parser.run()
3. assert expected.angle == oscillation.readings[0].angle

Reply:
I've implemented this in the test_oscillation_detected_up_down